Firestore/core/test/unit/bundle/bundle_reader_test.cc

```diff
@@ -0,0 +1,533 @@
+  AddDocument(Document1());
+
+  const auto& bundle =
+      BuildBundle("bundle-1", testutil::Version(6000004000), 1);
+  BundleReader reader(bundle_serializer,
+                      absl::make_unique<std::istringstream>(bundle));
+
+  std::vector<std::unique_ptr<BundleElement>> elements =
+      VerifyFullBundleParsed(reader, "bundle-1", testutil::Version(6000004000));
+
+  EXPECT_EQ(elements.size(), 2);
+  VerifyDocumentMetadataEquals(
+      *reinterpret_cast<BundledDocumentMetadata*>(elements[0].get()),
+      DocumentMetadata1());
+  VerifyDocumentEncodesToOriginal(
+      *reinterpret_cast<BundleDocument*>(elements[1].get()), Document1());
+}
+
+TEST_F(BundleReaderTest, ReadsWithDeletedDocument) {
+  AddDocumentMetadata(DeletedDocumentMetadata());
+  AddDocumentMetadata(DocumentMetadata2());
+  AddDocument(Document2());
+
+  const auto& bundle =
+      BuildBundle("bundle-1", testutil::Version(6000004000), 2);
+  BundleReader reader(bundle_serializer,
+                      absl::make_unique<std::istringstream>(bundle));
+
+  std::vector<std::unique_ptr<BundleElement>> elements =
+      VerifyFullBundleParsed(reader, "bundle-1", testutil::Version(6000004000));
+
+  EXPECT_EQ(elements.size(), 3);
+  VerifyDocumentMetadataEquals(
+      *reinterpret_cast<BundledDocumentMetadata*>(elements[0].get()),
+      DeletedDocumentMetadata());
+  VerifyDocumentMetadataEquals(
+      *reinterpret_cast<BundledDocumentMetadata*>(elements[1].get()),
+      DocumentMetadata2());
+  VerifyDocumentEncodesToOriginal(
+      *reinterpret_cast<BundleDocument*>(elements[2].get()), Document2());
+}
+
+TEST_F(BundleReaderTest, ReadsWithoutDocumentOrQuery) {
+  const auto& bundle =
+      BuildBundle("bundle-1", testutil::Version(6000004000), 0);
+  BundleReader reader(bundle_serializer,
+                      absl::make_unique<std::istringstream>(bundle));
+
+  std::vector<std::unique_ptr<BundleElement>> elements =
+      VerifyFullBundleParsed(reader, "bundle-1", testutil::Version(6000004000));
+
+  EXPECT_EQ(elements.size(), 0);
+}
+
+TEST_F(BundleReaderTest, ReadsLargeDocument) {
+  AddDocumentMetadata(DocumentMetadata2());
+  AddDocument(LargeDocument2());
+
+  const auto& bundle =
+      BuildBundle("bundle-1", testutil::Version(6000004000), 0);
+  std::cout << bundle << "\n";
+  BundleReader reader(bundle_serializer,
+                      absl::make_unique<std::istringstream>(bundle));
+
+  std::vector<std::unique_ptr<BundleElement>> elements =
+      VerifyFullBundleParsed(reader, "bundle-1", testutil::Version(6000004000));
+
+  VerifyDocumentMetadataEquals(
+      *reinterpret_cast<BundledDocumentMetadata*>(elements[0].get()),
+      DocumentMetadata2());
+  VerifyDocumentEncodesToOriginal(
+      *reinterpret_cast<BundleDocument*>(elements[1].get()), LargeDocument2());
+}
+
+TEST_F(BundleReaderTest, FailsWithBadLengthPrefix) {
+  const auto& bundle =
+      BuildBundle("bundle-1", testutil::Version(6000004000), 0);
+  for (int l = 1; l < 4; l++) {
```

Comment:
Nit: the style guide has recently been updated to require the prefix form of increment to be used in cases like this (see [here](go/cstyle#Preincrement_and_Predecrement)).